File: core/core-api/src/main/java/com/coffee/gu/stamp/StampService.java
package com.coffee.gu.stamp;

import com.coffee.gu.Principal;
import org.springframework.stereotype.Service;

import java.time.LocalDateTime;
import java.util.List;

@Service
public class StampService {

    private final StampReader stampReader;

    public StampService(StampReader stampReader) {
        this.stampReader = stampReader;
    }

    public StampCount count(Principal principal) {
        Long count = stampReader.getAvailableStampCounts(principal);
        return new StampCount(principal.getKey(), count);
    }

    public StampExpiringSoonCount countExpiringSoon(Principal principal) {
        LocalDateTime now = LocalDateTime.now();
        Long expiringSoonCount = stampReader.getExpiringSoonCount(principal, now);
        return new StampExpiringSoonCount(principal.getKey(), expiringSoonCount);
    }

    public List<StampHistory> getStampHistories(Principal principal) {
        return stampReader.getStampHistories(principal);
    }

}

File: core/core-api/src/main/kotlin/com/coffee/gu/stamp/StampService.kt
package com.coffee.gu.stamp

import com.coffee.gu.Principal
import org.springframework.stereotype.Service
import java.time.LocalDateTime

@Service
class StampService(
    private val stampReader: StampReader,
) {
    fun count(principal: Principal): StampCount {
        val count = stampReader.getAvailableStampCounts(principal)
        return StampCount(principal.key, count)
    }

    fun countExpiringSoon(principal: Principal): StampExpiringSoonCount {
        val now = LocalDateTime.now()
        val expiringSoonCount = stampReader.getExpiringSoonCount(principal, now)
        return StampExpiringSoonCount(principal.key, expiringSoonCount)
    }

    fun getStampHistories(principal: Principal): List<StampHistory> {
        return stampReader.getStampHistories(principal)
    }
}
